pandas/core/generic.py

```diff
@@ -7367,16 +7367,124 @@ def _clip_with_one_bound(self, threshold, method, axis, inplace):
                 threshold = align_method_FRAME(self, threshold, axis, flex=None)[1]
         return self.where(subset, threshold, axis=axis, inplace=inplace)
 
+    @overload
+    def clip(
+        self: FrameOrSeries,
+        lower=...,
+        upper=...,
+        axis: Axis | None = ...,
```

Comment:
users of the pandas api should not pass None.

```
axis: int or str axis name, optional
```

for the default (should be listed in the docstring instead of optional - not for this PR though), the argument should not be passed.

@MarcoGorelli what are you thoughts about not including `None` in the overloads and also now that we are using PEP 604, removing `no_implicit_optional = True` from `setup.cfg` and excluding `None` actual function signatures too?

Reply:
Hey @simonjayhawkins 

+1 on removing `no_implicit_optional`

Not sure what you mean about not including `None` in overloads - if `axis=None` is passed internally, then won't it need to be in one of the overloads for type checking to work?

Reply:
I guess the question is (in general) should `None` ever be passed internally or should kwargs always be (in general) validated at the top level.

Reply:
IIRC we have had issues reported where `None` is passed by users and when we change defaults (I think bool True/False in a specific case I remember recently) this is results in a perceived regression. We also sometime use `lib.nodefault` to detect if a kwarg has been specified. So I was thinking to that discouraging users passing `None` could be worthwhile.

Reply:
@simonjayhawkins I looked into the `no_implicit_optional = True`, but it seems it's only possible to omit if you only have two possible types.

e.g. the following works:
```python
from __future__ import annotations
from typing import Tuple

def foo(
    a: str | int | None  = None,
    b: str = None,    
): ...
```

but the following doesn't:
```python
from __future__ import annotations
from typing import Tuple

def foo(
    a: str | int  = None,
    b: str = None,    
): ...
```
, we get
```
t.py:5: error: Incompatible default for argument "a" (default has type "None", argument has type "Union[str, int]")  [assignment]
Found 1 error in 1 file (checked 1 source file)
```

So I think it's cleaerer to keep `no_implicit_optional`